From d6492c4376abcd05a004e38a7834272ceb49e288 Mon Sep 17 00:00:00 2001
From: Debbie Martin <Debbie.Martin@arm.com>
Date: Tue, 15 Oct 2024 18:21:00 +0100
Subject: [PATCH] transport: Handle SCMI payload size error

The transport module is responsible for managing the SCMI message header
functionality, which includes checking the size of the command payload and
copying the payload. If the reported size exceeds the command buffer size,
the platform should respond with an error and exit the function without further
command processing. However, the current implementation checks the size and
reports the error but fails to exit the function. Instead, it continues copying
and processing the command, leading to a buffer overflow by overwriting the SCP
internal memory. This change will ensure the command handling terminates
correctly by exiting the function.

This change addresses the security vulnerability (CVE-2024-9413).

Signed-off-by: Tarek El-Sherbiny <tarek.el-sherbiny@arm.com>
Upstream-Status: Backport [https://gitlab.arm.com/firmware/SCP-firmware/-/commit/87a240ae998f694d1d1a87cfed33a9853afb6019]
Signed-off-by: Debbie Martin <Debbie.Martin@arm.com>
---
 module/transport/src/mod_transport.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/module/transport/src/mod_transport.c b/module/transport/src/mod_transport.c
index 662a6c027..733548cdb 100644
--- a/module/transport/src/mod_transport.c
+++ b/module/transport/src/mod_transport.c
@@ -599,6 +599,8 @@ static int transport_message_handler(struct transport_channel_ctx *channel_ctx)
                 channel_ctx->transport_signal.firmware_signal_api->signal_error(
                     channel_ctx->service_id);
         }
+
+        return status;
     }

     if (transport_type == MOD_TRANSPORT_CHANNEL_TRANSPORT_TYPE_OUT_BAND) {
--
2.25.1

